frontend/src/metabase/visualizations/lib/trends.js

```diff
@@ -3,34 +3,35 @@ import moment from "moment-timezone";
 
 // mappings of allowed operators
 const EXPRESSION_OPERATORS = new Map([
-  ["+", (...args) => `(${args.join(" + ")})`],
-  ["-", (...args) => `(${args.join(" - ")})`],
-  ["*", (...args) => `(${args.join(" * ")})`],
-  ["/", (...args) => `(${args.join(" / ")})`],
-  ["log", x => `Math.log(${x})`],
-  ["pow", (x, y) => `Math.pow(${x}, ${y})`],
-  ["exp", x => `Math.pow(Math.E, ${x})`],
+  ["+", (...args) => args.reduce((x, y) => x + y)],
+  ["-", (...args) => args.reduce((x, y) => x - y)],
+  ["*", (...args) => args.reduce((x, y) => x * y)],
+  ["/", (...args) => args.reduce((x, y) => x / y)],
+  ["log", x => Math.log(x)],
+  ["pow", (x, y) => Math.pow(x, y)],
+  ["exp", x => Math.pow(Math.E, x)],
 ]);
 // list of allowed expressions
 const EXPRESSION_IDENTIFIERS = new Set(["x"]);
 
-function compileNode(node) {
+function computeExpression(node, x) {
   if (Array.isArray(node)) {
     const [operator, ...args] = node;
     if (EXPRESSION_OPERATORS.has(operator)) {
-      return EXPRESSION_OPERATORS.get(operator)(...args.map(compileNode));
+      const operatorFn = EXPRESSION_OPERATORS.get(operator);
+      const argValues = args.map(arg => computeExpression(arg, x));
+      return operatorFn(...argValues);
     }
   } else if (typeof node === "number") {
     return node;
   } else if (typeof node === "string" && EXPRESSION_IDENTIFIERS.has(node)) {
-    return node;
+    return x;
   }
   throw new Error(`Invalid expression: ${node}`);
 }
 
 export function compileExpression(node) {
-  const compiled = compileNode(node);
-  return new Function("x", `return ${compiled};`);
+  return x => computeExpression(node, x);
```

Comment:
I think we don't need `compileExpression` now, we can just inline it.

There is just one usage in a conditional statement. The `else` block already defines a function inline, so it'll go along nicely.

```diff
--- a/frontend/src/metabase/visualizations/lib/trends.js
+++ b/frontend/src/metabase/visualizations/lib/trends.js
export function getTrendDataPointsFromInsight(insight, xDomain, count = 10) {
 
   let fn;
   if (insight["best-fit"]) {
-    fn = compileExpression(insight["best-fit"]);
+    fn = x => computeExpression(insight["best-fit"], x);
   } else {
     fn = x => x * insight.slope + insight.offset;
   }
```

Reply:
Agree, fixed